homeassistant/components/osramlightify/light.py

```diff
@@ -4,270 +4,410 @@
+        return self._effect_list
+
+    @property
+    def min_mireds(self):
+        """Return the coldest color_temp that this light supports."""
+        return self._min_mireds
+
+    @property
+    def max_mireds(self):
+        """Return the warmest color_temp that this light supports."""
+        return self._max_mireds
+
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def state_attributes(self):
+        """Return optional state attributes."""
+        data = {}
+        if self._supported_features & SUPPORT_COLOR_TEMP:
+            data[ATTR_MIN_MIREDS] = self._min_mireds
+            data[ATTR_MAX_MIREDS] = self._max_mireds
+
+        if self._supported_features & SUPPORT_EFFECT:
+            data[ATTR_EFFECT_LIST] = self._effect_list
+
+        if self.is_on:
+            if self._supported_features & SUPPORT_BRIGHTNESS:
+                data[ATTR_BRIGHTNESS] = self._brightness
+
+            if self._supported_features & SUPPORT_COLOR_TEMP:
+                data[ATTR_COLOR_TEMP] = self._color_temp
+
+            if self._supported_features & SUPPORT_COLOR:
+                data[ATTR_RGB_COLOR] = self._rgb_color
+                data[ATTR_HS_COLOR] = self._hs_color
+                data[ATTR_XY_COLOR] = self._xy_color
+
+        return data
+
+    def play_effect(self, effect, transition):
+        """Play selected effect."""
+        if effect == EFFECT_RANDOM:
+            self._rgb_color = (random.randrange(0, 256),
+                               random.randrange(0, 256),
+                               random.randrange(0, 256))
+            self._hs_color = color_util.color_RGB_to_hs(*self._rgb_color)
+            self._xy_color = color_util.color_RGB_to_xy(*self._rgb_color)
+            self._luminary.set_rgb(*self._rgb_color, transition)
+            self._luminary.set_onoff(True)
+            return True
+
+        return False
 
     def turn_on(self, **kwargs):
         """Turn the device on."""
-        if ATTR_TRANSITION in kwargs:
-            transition = int(kwargs[ATTR_TRANSITION] * 10)
-        else:
-            transition = 0
+        transition = int(kwargs.get(ATTR_TRANSITION, 0) * 10)
+        if ATTR_EFFECT in kwargs:
+            if self.play_effect(kwargs[ATTR_EFFECT], transition):
+                self.schedule_update_ha_state()
 
-        if ATTR_BRIGHTNESS in kwargs:
-            self._brightness = kwargs[ATTR_BRIGHTNESS]
-            self._luminary.set_luminance(
-                int(self._brightness / 2.55), transition)
-        else:
-            self._luminary.set_onoff(1)
+            return
 
         if ATTR_HS_COLOR in kwargs:
-            red, green, blue = \
-                color_util.color_hs_to_RGB(*kwargs[ATTR_HS_COLOR])
-            self._luminary.set_rgb(red, green, blue, transition)
+            self._hs_color = kwargs[ATTR_HS_COLOR]
+            self._rgb_color = color_util.color_hs_to_RGB(*self._hs_color)
```

Comment:
Do not set `_rgb_color` here, your `update()` will be called right after you return from this method.

Reply:
Yes, I know.  But `update()` will need to get an answer from the physical device first that may take up to a second (or even more in some rare cases).  Updating all the properties immediately inside `turn_on()/turn_off()` (first of all, `self._is_on`) provides a more smooth experience (I have tested both options in practice).  Does it make sense?

Reply:
What is the issue that you see by letting it take a second?

Reply:
The next `state_attributes` may be called before `update` finished.  I have `scan_interval` set to 1 second.